caiman/source_extraction/cnmf/utilities.py

```diff
@@ -45,7 +41,163 @@
+                output += cval
+                return output                
+        return cv2.GaussianBlur(input, ksize=ksize[::-1], sigmaX=sigma[1], sigmaY=sigma[0], dst=output,
+                                borderType=borderType)
+    
+    elif input.ndim==3:
+        if mode=='constant' and cval!=0:
+            input = input-cval 
+        if output is None:
+            output = np.zeros_like(input)
+        for i in range(input.shape[1]):
+            output[:,i] = cv2.GaussianBlur(input[:,i], ksize=ksize[2::-2],
+                                         sigmaX=sigma[2], sigmaY=sigma[0], borderType=borderType)
+        if np.isfortran(input):
+            for i in range(input.shape[2]):
+                output[...,i] = cv2.GaussianBlur(output[...,i], ksize=(ksize[1],1),
+                                             sigmaX=sigma[1], sigmaY=0, borderType=borderType)    
+        else:
+            for i in range(input.shape[0]):
+                output[i] = cv2.GaussianBlur(output[i], ksize=(1,ksize[1]),
+                                             sigmaY=sigma[1], sigmaX=0, borderType=borderType)
+        if mode=='constant' and cval!=0:
+            output+=cval
+        return output
+
+
+def uniform_filter(input, size=3, output=None, mode='reflect', cval=0.0):
+    """Multidimensional uniform filter.
+    A faster version of scipy.ndimage.uniform_filter
+    Uses OpenCV for 2 and 3 dimensional input, otherwise scipy
+    """
+    if input.ndim not in (2,3):
+        return scipy.ndimage.uniform_filter(input, size, output, mode, cval)
+    size = np.atleast_1d(size).tolist()
+    if len(size)==1:
+        size = size*input.ndim
+    borderType = get_border_type(mode)
+
+    if input.ndim==2:
+        if mode=='constant' and cval!=0:
+            if output is None:
+                return cval+cv2.boxFilter(input-cval, -1, ksize=size[::-1], borderType=borderType)
+            else:
+                cv2.boxFilter(input-cval, -1, ksize=size[::-1], dst=output, borderType=borderType)
+                output += cval
+                return output
+        return cv2.boxFilter(input, -1, ksize=size[::-1], dst=output, borderType=borderType)
+
+    elif input.ndim==3:
+        if mode=='constant' and cval!=0:
+            input = input-cval
+        if output is None:
+            output = np.zeros_like(input)
+        for i in range(input.shape[1]):
+            output[:,i] = cv2.boxFilter(input[:,i], -1, ksize=size[2::-2], borderType=borderType)
+        if np.isfortran(input):
+            for i in range(input.shape[2]):
+                output[...,i] = cv2.boxFilter(output[...,i], -1, ksize=(size[1],1), borderType=borderType)
+        else:
+            for i in range(input.shape[0]):
+                output[i] = cv2.boxFilter(output[i], -1, ksize=(1,size[1]), borderType=borderType)
+        if mode=='constant' and cval!=0:
+            output+=cval
+        return output
+
```

Comment:
Should we have an else at the end that throws an exception if ndim is somehow some other value? Feels weird to fall out of the function if the latter elif is false